linera-core/src/chain_worker/state.rs

```diff
@@ -975,13 +1189,27 @@ where
         // That means that when the write-lock is acquired, no readers will be waiting to acquire
         // the lock. This is important because otherwise readers could have a stale view of the
         // chain state.
-        let maybe_shared_chain_view = self.shared_chain_view.take();
+        let maybe_shared_chain_view = self.state.shared_chain_view.take();
         let _maybe_write_guard = match &maybe_shared_chain_view {
             Some(shared_chain_view) => Some(shared_chain_view.write().await),
             None => None,
         };
```

Comment:
I'm a bit confused about this — the write guard doesn't seem to be used?

Reply:
In order to allow staging changes to memory concurrently, there needs to be two copies. One stays with `ChainWorkerState`, and the other is sent to the requester, wrapped inside a read-lock. However, to ensure that the requester doesn't end up with incorrect information, changes are only persisted to disk _after_ the requester has finished using the read-lock. The way this is done is by acquiring the write-lock, but instead of using its contents (which is a copy of the `ChainStateView` before changes were staged), use the original `ChainStateView` that has the staged changes by the `ChainWorkerState`.

Reply:
Thanks for the explanation!  This feels a little error-prone — but it wasn't changed as part of this PR so not blocking on it :)